### PR TITLE
Make the new operator protected for some IR types.

### PR DESCRIPTION
**Pull request by fruffy**
Addresses some of the points raised in #4612. Note that #4612 is still required. Type pointers can still not be unique because of the source information associated with the type and the `clone` call.

### CODE REVIEW

**Comment by vlstill** [comment]
> Addresses some of the points raised in #4612. Note that #4612 is still required. **Type pointers can still be unique** because of the source information associated with the type and the `clone` call.

Did you meant to say the type pointers can't be unique?

**Comment by fruffy** [comment]
> > Addresses some of the points raised in #4612. Note that #4612 is still required. **Type pointers can still be unique** because of the source information associated with the type and the `clone` call.
> 
> Did you meant to say the type pointers can't be unique?

Yup

File: ir/base.def [hunk 1]
```diff
@@ -120,6 +120,7 @@ abstract Type_Base : Type {
 class Type_Unknown : Type_Base {
 #nodbprint
     static Type_Unknown get();
+    static Type_Unknown get(const Util::SourceInfo &si);
```
Comment (ChrisDodd) [inline ir/base.def:123]: Is there ever a case where we have an unknown type with a source info?  I can't think of how it could occur.

**Review by ChrisDodd — APPROVED** [review]

File: ir/type.cpp [hunk 2]
```diff
@@ -87,41 +88,65 @@ const Type_Bits *Type_Bits::get(int width, bool isSigned) {
     return result;
 }
 
-const Type::Unknown *Type::Unknown::get() {
-    static const Type::Unknown *singleton = nullptr;
-    if (!singleton) singleton = (new Type::Unknown());
+const Type_Bits *Type_Bits::get(const Util::SourceInfo &si, int sz, bool isSigned) {
+    auto *result = new IR::Type_Bits(si, sz, isSigned);
+    if (sz < 0) {
+        ::error(ErrorType::ERR_INVALID, "%1%: Width of type cannot be negative", result);
```
Comment (asl) [inline ir/type.cpp:94]: Should this be `BUG_CHECK` instead? Also, it seems we're creating `result` here just for some message printing. I think we can just use `sz` and create result only when all sanity checks pass

File: ir/base.def (hunk 1 above)
Comment (fruffy) [reply ir/base.def:123]: I do not think so. At least we do not store any source info for an unknown type when parsing. I mostly mirrored what the auto-generated constructors allow, so I do not think at can hurt at least.

File: ir/type.cpp (hunk 2 above)
Comment (fruffy) [reply ir/type.cpp:94]: I am not sure why an error was chosen here but the comment 
```
// Return a value that will not cause crashes later on
```
indicates this was deliberate? 

I moved the result allocation down and changed the wording a bit.

**Pull request merged**